### ISSUE TITLE
Customizations for generated code

### ISSUE THREAD

**Issue by MahdiBM** [apple/swift-openapi-generator#405]
In a project we en/decode our enums in a custom way, so much that we've recently created a macro for it called `CodableEnum`. This sort of customization seems not possible with the current OpenAPI generator.

I'm thinking the generator could take names of macros to attach to certain declarations in the generator config file, which would provide maximum flexibility around these stuff. This would be really nice to have, but is not as easy to provide as it sounds like:

* There will need to be a way for users to disable generation of certain things like the Decodable and Encodable conformances, so that they can provide those using their own macro, if needs be.
* This has source-compatibility implications because if the generator adds a new property to a kind of types, and a macro already provides a property with the same name, there will be errors. 
So the source-compatibility can only be guaranteed in patch versions, and not minors.

**Comment by MahdiBM** [comment]
For `CodableEnum`, i think an example would be the most efficient way of describing what it does:
```swift
TestCodableEnum.extendedUser(
    id: 1,
    fullName: "Mahdi Bahrami",
    isGoodUser: true
)
```
is encoded to:
```json
{
  "type": "extended_user", 
  "id": 1,
  "full_name": "Mahdi Bahrami",
  "is_good_user": true
}
```
* `type` is reserve for the enum case name (`extendedUser`).
* `snake_case` for keys.

**Comment by MahdiBM** [comment]
> Now, changing the generation logic so that the codable implementations aren't generated is a complication, this would suddenly mean that we'd have to understand that some macros specifically provide custom Codable implementations. So let's first see if there's a way we could avoid that.

Right, optimally we should avoid that. But generally the generator can take arguments about whether or not to generate some special property/function for the types too, the same way it accepts attaching macros.

**Comment by czechboy0** [comment]
> For `CodableEnum`, i think an example would be the most efficient way of describing what it does:
> 
> ```swift
> TestCodableEnum.extendedUser(
>     id: 1,
>     fullName: "Mahdi Bahrami",
>     isGoodUser: true
> )
> ```
> 
> is encoded to:
> 
> ```json
> {
>   "type": "extended_user", 
>   "id": 1,
>   "full_name": "Mahdi Bahrami",
>   "is_good_user": true
> }
> ```
> 
>  * `type` is reserved for the enum case name (`extendedUser`).
> * `snake_case` for keys.

That looks a lot like a `oneOf` with a discriminator, which we do support:


Your `#/components/schemas` would look like this:

```yaml
ExtendedUser:
  type: object
  properties:
    type:
      type: string
    id:
      type: integer
    full_name:
      type: string
    is_good_user:
      type: bool
  required:
    - type
    - id
    - full_name
    - is_good_user
TestCodableEnum:
  oneOf:
    - $ref: '#/components/schemas/ExtendedUser'
  discriminator:
    propertyName: type
    mapping:
      extendedUser: '#/components/schemas/ExtendedUser'
```

One example from our integration tests: https://github.com/apple/swift-openapi-generator/blob/e30691e5d3613994b04b142df5c52158d066b0bb/Tests/OpenAPIGeneratorReferenceTests/Resources/Docs/petstore.yaml#L464C7-L464C7

**Comment by czechboy0** [comment]
Not really, we need to keep things working for even the more complex cases, we don't generally flatten namespaces when they're not needed for two reasons. 
1. It's very complicated to know for sure that you can flatten a namespace.
2. It makes API evolution painful for adopters, as adding a second element to an array shouldn't suddenly change the whole nature of a type (e.g. add a new scope).

The generated code is the type-safe base layer that folks should build conveniences on. But macros could allow to make that pretty easy to do, if you wrote a macro that verified that a function/type is simple enough for you to generate some convenience accessors on it, that would make sense to me.

As long as the only thing the generator has to do is to attach `@Foo` on a type or an operation function, it's realistic that this feature could be added, and without breaking API (which is important at this stage).

**Comment by simonjbeaumont** [comment]
Yes, I would agree. It’s reasonable for us to think about adding custom macros, but I don’t think we should try to be any smarter than that. 

Could be a post-v1 feature as I don’t foresee any API breaking changes needed.

**Comment by czechboy0** [comment]
Ok now that we better understand your ask, it does seem like a dup of #383, so let's centralize further discussion and design there.

**Comment by MahdiBM** [comment]
> It's very complicated to know for sure that you can flatten a namespace.

I see ... you're right. I guess the macro _could_ already provide those as static functions if I really wanted to stick with them? Not 100% sure. Either way this is good enough to me although it's not optimal 🤔.

**Comment by czechboy0** [comment]
Naming improvements landed in 1.6.0 as the new idiomatic naming strategy: https://swiftpackageindex.com/apple/swift-openapi-generator/main/documentation/swift-openapi-generator/soar-0013
